### PR TITLE
WIP - crude first pass at node #return functionality

### PR DESCRIPTION
**Pull request by rcuddy**
Re: https://github.com/thesecretlab/YarnSpinner/issues/92
This works in my limited test case used so far, but I expect a) many edge cases and b) I've probably missed something really obvious that someone who understands this project should cast an eye over.

I've not begun testing with variables, to see what happens on state replay,  but vaguely functional flow control at least. PR being created for comment from the YS guys who can hopefully advise if I'm going in the wrong direction or not.

```
title: Start
tags: tag1
colorID: 0
position: 0,0
---
Start: This is a line in start block. #line:a76b29
Start: This is too, in the start block. #line:7b5c3c
[[JumpReturn1]] #return
Start: And we're back in start block, at the line after the jump! #line:0c8b0c
[[JumpReturn2]] #return
Start: And we're back in start block, and next we'll test menus;
->Go and come back?
    [[QuickChoiceReturn]] #return
-> Or stay in this block?
    We stayed put...
And now the game is going to end with a normal end of block jump.
[[End]]
===
title: JumpReturn1
tags: tag2
colorID: 0
position: 0,0
---
JumpReturn1: Line 3 is froma jump, and we're now in a new block. #line:4d4894
JumpReturn1: Line 4 is the next line in the new block. #line:b26d94
===
title: JumpReturn2
tags: tag3
colorID: 0
position: 0,0
---
JumpReturn2: Line 5 is jumping back into jumpblock 2! #line:7daa33
JumpReturn2: Line 6 is also in jumpblock 2. #line:88a7f3
===
title: End
tags: tag4
colorID: 0
position: 0,0
---
End: We ran out of calls and went to the End block. Game over! #line:a2e4d5
===
title: QuickChoiceReturn
---
We jumped in a menu!
===
```

Results in expected output:

```
YarnSpinnerConsole.exe run ../../test.yarn.txt
Start: This is a line in start block.
Start: This is too, in the start block.
JumpReturn1: Line 3 is froma jump, and we're now in a new block.
JumpReturn1: Line 4 is the next line in the new block.
Start: And we're back in start block, at the line after the jump!
JumpReturn2: Line 5 is jumping back into jumpblock 2!
JumpReturn2: Line 6 is also in jumpblock 2.
Start: And we're back in start block, and next we'll test menus;
Options:
1. Go and come back?
2.  Or stay in this block?
? 1
We jumped in a menu!
And now the game is going to end with a normal end of block jump.
End: We ran out of calls and went to the End block. Game over!
```
or selecting 2 for our choice winds up continueing on as expected.

### CODE REVIEW

File: YarnSpinner/VirtualMachine.cs [hunk 1]
```diff
@@ -328,10 +336,22 @@ public enum ExecutionState {
                 /// - Stop
                 /** Immediately stop execution, and report that fact.
                  */
-                nodeCompleteHandler (new Dialogue.NodeCompleteResult (null));
-                executionState = ExecutionState.Stopped;
 
+                if (_returnStates.Count > 0)
```
Comment (rcuddy) [inline YarnSpinner/VirtualMachine.cs]: I'm conscious of the fact that this is redefining the meaning of "stop" but I'm not sure where or how else to have injected this unwind. Happy for suggestions from on high,

File: YarnSpinner/VirtualMachine.cs [hunk 2]
```diff
@@ -328,10 +336,22 @@ public enum ExecutionState {
                 /// - Stop
                 /** Immediately stop execution, and report that fact.
                  */
```
Comment (rcuddy) [inline YarnSpinner/VirtualMachine.cs]: I'm conscious of the fact that this is redefining the meaning of "stop" but I'm not sure where or how else to have injected this unwind. Happy for suggestions from on high,

**Comment by rcuddy** [comment]
Okay further testing done; definitely still needs work.

Doing a #return inside of another #return does not unwind properly.

**Comment by rcuddy** [comment]
Unwinding multiple levels now works!

```
title: Start
tags: tag1
colorID: 0
position: 0,0
---
Start: This is a line in start block. #line:a76b29
Start: This is too, in the start block. #line:7b5c3c
[[JumpReturn1]] #return
Start: And we're back in start block, at the line after the jump! #line:0c8b0c
Start: And we're back in start block, and next we'll test menus;
->Go and come back?
    [[QuickChoiceReturn]] #return
-> Or stay in this block?
    We stayed put...
And now the game is going to end with a normal end of block jump.
[[End]]
===
title: JumpReturn1
tags: tag2
colorID: 0
position: 0,0
---
JumpReturn1: Line 3 is froma jump, and we're now in a new block. #line:4d4894
JumpReturn1: Line 4 is the next line in the new block. #line:b26d94
[[JumpReturn2]] #return
===
title: JumpReturn2
tags: tag3
colorID: 0
position: 0,0
---
JumpReturn2: Line 5 is jumping back into jumpblock 2! #line:7daa33
JumpReturn2: Line 6 is also in jumpblock 2. #line:88a7f3
[[JumpReturn3]]
===
title: End
tags: tag4
colorID: 0
position: 0,0
---
End: We ran out of calls and went to the End block. Game over! #line:a2e4d5
===
title: QuickChoiceReturn
---
We jumped in a menu!
===
title: JumpReturn3
tags: tag3
colorID: 0
position: 0,0
---
JumpReturn3: Line 7 is jumping back into jumpblock 3!
JumpReturn3: Line 8 is also in jumpblock 3. 
```

**Comment by rcuddy** [comment]
Furthermore; I've now finished testing with variable changes as we jump / unwind and it seems okay.

This set of changes is now exactly what I needed for my own purposes, but please, guardians of the code base, let me know if I need to do anything better or different.

**Comment by rcuddy** [comment]
Will fix the newly created conflict tonight.

**Comment by desplesda** [comment]
Awesome!

It's worth noting that we'll probably want to include this in the new compiler (see AntlrCompiler.cs), as well.

**Comment by rcuddy** [comment]
Ah, okay, will have a look! Might take a bit longer than tonight then depending on how different the AnltrCompiler is! Thanks for the heads up re that file.

**Comment by desplesda** [comment]
No problem! Sorry for springing that on you!

**Comment by rcuddy** [comment]
Okay well the positive is that this still works, I just had to move the ByteCode to its new home.

The negative is that i have no idea how antlr works and will need to do some reading about how to actually run that  mode so i can step through properly and make sure this works there too.

Also,  I'm not happy with my redefining of the ByteCode.Stop instruction to allow an unwind. I think I might make return Stop to killing the stack completely as per existing functionality and make a new call to force the unwind without killing execution 100%.

So I'll update this over the next few days with these changes hopefully.
